src/Microsoft.ML.Data/Transforms/HashTransform.cs

```diff
@@ -523,6 +549,11 @@ private ValueGetter<VBuffer<uint>> ComposeGetterVec(IRow input, int iinfo)
                 return ComposeGetterVecCore<ushort>(input, iinfo, HashUnord, HashDense, HashSparse);
             case DataKind.U4:
                 return ComposeGetterVecCore<uint>(input, iinfo, HashUnord, HashDense, HashSparse);
+            case DataKind.R4:
+                // Ignoring HashSparse for I1-R8, since 0 will be hashed to a non zero value.
+                return ComposeGetterVecCore<Single>(input, iinfo, HashUnord, HashDense, HashDense);
+            case DataKind.R8:
+                return ComposeGetterVecCore<Double>(input, iinfo, HashUnord, HashDense, HashDense);
```

Comment:
>ComposeGetterVecCore [](start = 23, length = 20)

This will not work. This utility function *assumes* that the hashing is sparsity preserving, and you will note that in your implementation of `HashDense` you are nowhere filling in the value of the 0 hash anywhere.

So: let's imagine that the hashes of 0.0, 0.2, 0.4 were, say, 5, 31, 7, respectively. Then we would expect an input source `VBuffer<double>` of length 4with sparse values { 0:0.2, 3:0.4 }, to result in the output vector { 31, 5, 5, 7 }, but this code will instead result in { 31, 0, 0, 7 }.

The issue is sufficiently complex that you would benefit from adding a test to ensure that the outputs are real values.

Reply:
This is still problematic in the latest commit, do you have a plan to address?

---
In reply to: [210720336](https://github.com/dotnet/machinelearning/pull/684#discussion_r210720336) [](ancestors = 210720336)